iOSDFULibrary/Classes/Implementation/SecureDFU/Services/SecureDFUService.swift

```diff
@@ -394,45 +394,66 @@ import CoreBluetooth
+                    // devices with bond information sharing between app and the bootloader.
+                    else if uuidHelper.matchesButtonless(characteristic) {
+                        buttonlessDfuCharacteristic = ButtonlessDFU(characteristic, logger)
+                        buttonlessDfuCharacteristic?.uuidHelper = uuidHelper
+                        _success?()
+                        return
+                    }
+                    // End
                 }
-                // Support for Buttonless DFU Service from SDK 12.x (as experimental).
-                // SDK 13 added a new characteristic in Secure DFU Service with buttonless
-                // feature without bond sharing (bootloader uses different device address).
-                // SDK 14 will add a new characteristic with buttonless service for bonded
-                // devices with bond information sharing between app and the bootloader.
-                else if uuidHelper.matchesButtonless(characteristic) {
-                    buttonlessDfuCharacteristic = ButtonlessDFU(characteristic, logger)
-                    buttonlessDfuCharacteristic?.uuidHelper = uuidHelper
-                    _success?()
-                    return
+            }
+            
+            // Log what was found in case of an error
+            if dfuPacketCharacteristic == nil || dfuControlPointCharacteristic == nil {
+                if let characteristics = service.characteristics, characteristics.isEmpty == false {
+                    logger.d("The following characteristics were found:")
+                    characteristics.forEach { characteristic in
+                        logger.d(" - \(characteristic.uuid.uuidString)")
+                    }
+                } else {
+                    logger.d("No characteristics found in the service")
                 }
-                // End
+                logger.d("Is it the right device? If so, it may be caching issue. Try again after restarting Bluetooth. Make sure that your device has the Service Changed characteristic")
             }
             
             // Some validation
             if dfuControlPointCharacteristic == nil {
-                logger.e("DFU Control Point characteristics not found")
+                logger.e("DFU Control Point characteristic not found")
                 // DFU Control Point characteristic is required
                 _report?(.deviceNotSupported, "DFU Control Point characteristic not found")
                 return
             }
             if dfuPacketCharacteristic == nil {
-                logger.e("DFU Packet characteristics not found")
+                logger.e("DFU Packet characteristic not found")
                 // DFU Packet characteristic is required
                 _report?(.deviceNotSupported, "DFU Packet characteristic not found")
                 return
             }
             if !dfuControlPointCharacteristic!.valid {
```

Comment:
This is an older piece of code, but maybe it can be improved ? `!dfuControlPointCharacteristic!` looks a bit funny. 😆 but that's just a comment and it dosen't need to be changed for now

Reply:
Yeah... those ! at the beginning are not readable. I would really like to see `not` operator instead.

Reply:
#define ! not 😆

Reply:
Make it `value == false` instead.